### PR TITLE
Add multiple runs eval

### PR DESCRIPTION
**Pull request by binkjakub**
# CHANGELOG
1. Fixed computation of `chrf` metric - previously, the results were overestimated since `chrf ` metric was wrongly. computed. We fixed it and provide correct metrics. Now metrics are slightly lower, however, we can still observe significant improvement after supervised fine-tuning.
3. Fixed `attention_mask` bug - missing `attention_mask` during inference was fixed and now inference is correct and can be done in batches 
4. Adjusted hyperparameters - we slightly adjusted hyperparameters based on several scientific papers and practice.
5. Added multiple inference runs to determine error bars - now we make inference with sampling and `temperature=0.1` ,  and perform inference 3 times with different random seeds; results tables have standard deviation included now.
7. Added evaluation with `llm-as-judge` approach - we utilize [`llama-3.1-70B-Instruct`](https://huggingface.co/meta-llama/Meta-Llama-3.1-8B-Instruct) and evaluated all models on that (code is ready to use and OpenAI-compatible endpoint).
8. Added [`mistralai/Mistral-Nemo-Instruct-2407`](https://huggingface.co/mistralai/Mistral-Nemo-Instruct-2407) LLM - previously used [`mistralai/Mistral-7B-Instruct-v0.3`](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3) LLM had problems with proper yaml formatting. We hypothesize that this might have been mitigated by carefully tuning prompt, however, decided to switch to newer model as it turned out to work better with the same prompt as Lllama-3.
9. Added [`speakleash/Bielik-7B-Instruct-v0.1`](https://huggingface.co/speakleash/Bielik-7B-Instruct-v0.1) model to include LLM pre-trained on Polish data,
10. Added small code(style) refinements and unit tests for the most crucial parts of code.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		pytorch \
+		cudatoolkit \
+		'xformers<0.0.27' \
+		-c pytorch \
+		-c nvidia \
+		-c xformers \
+		--yes
+	pip install "unsloth[colab-new] @ git+https://github.com/unslothai/unsloth.git"
+	pip install flash-attn --no-build-isolation
+	pip install -r requirements_unsloth.txt
```
Comment (asawczyn) [inline Makefile]: Why not moving this to the separate file? Conda utilizes `environment.yml` file https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-file-manually

File: README.md [hunk 2]
```diff
@@ -23,12 +23,80 @@ fostering cross-disciplinary and cross-jurisdictional collaboration.
 
 ![baner](https://raw.githubusercontent.com/pwr-ai/JuDDGES/bffb1d75ba7c78f101fc94bd9086499886b2c128/nbs/images/baner.png)
 
+## Usage
+
+### Installation
+
+- to install necessary dependencies use available `Makefile`, you can
+  use `python>=3.10`: `shell     make install`
```
Comment (asawczyn) [inline README.md]:
```suggestion
- to install necessary dependencies use available `Makefile`, use `python>=3.10`: `make install`
```
Moreover, I will recommend using one certain version of python as we didn't test every version i.e. 3.12

File: README.md [hunk 3]
```diff
@@ -23,12 +23,80 @@ fostering cross-disciplinary and cross-jurisdictional collaboration.
...
+- if you want to run evaluation and fine-tuning with `unsloth`, use the
+  following command with `python=3.10` inside conda environment:
+  `shell     make install_unsloth`
```
Comment (asawczyn) [inline README.md]:
```suggestion
- to run evaluation and fine-tuning with `unsloth`, use the
  following command with `python=3.10` inside conda environment:
  `make install_unsloth`
```

File: README.md [hunk 4]
```diff
@@ -23,12 +23,80 @@ fostering cross-disciplinary and cross-jurisdictional collaboration.
...
+
+### Dataset creation
+
+The specific details of dataset creation are available in
+[scripts/README.md](scripts/README.md).
+
+### Inference, fine-tuning and evaluation
+All commands for running inference, fine-tuning, and evaluation are declared as stages in the [`dvc.yaml`](dvc.yaml) file (see [DVC docs for details](https://dvc.org/doc/user-guide)).
+Each stage is set up as a matrix, meaning it runs for a combination of different parameters (e.g. models and random seeds).
```
Comment (asawczyn) [inline README.md]: Some stages don't use matrix

File: README.md [hunk 5]
```diff
@@ -23,12 +23,80 @@ fostering cross-disciplinary and cross-jurisdictional collaboration.
...
+> To introduce a new model, either from `hf-hub` or a local model/adapter, add its configuration to the `configs/model` directory.
+
+> [!TIP]
+> To run a stage for a single combination of parameters from the DVC matrix, simply run it with its full name, e.g., `predict@Bielik-7B-Instruct-v0.1-42` (check for names with `dvc stage list <stage_name>`).
+
+#### Inference
+  - Configuration file: `configs/predict.yaml`
+  - Command:
+    ```bash
+    CUDA_VISIBLE_DEVICES=0 NUM_PROC=10 dvc repro predict
```
Comment (asawczyn) [inline README.md:65]: These envs are specific for each user. Maybe explain them?

File: dashboards/pages/02_🔍_Analyse_Extracted_Information.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+output = io.BytesIO()
+with pd.ExcelWriter(output, engine="xlsxwriter") as writer:
+    df.to_excel(writer, sheet_name="Sheet1", index=False)
+output.seek(0)
+st.download_button(
+    label="Download data as Excel",
+    data=output,
+    file_name="judgements.xlsx",
+    mime="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
+)
```
Comment (asawczyn) [inline dashboards/pages/02_🔍_Analyse_Extracted_Information.py:51]: I'm not a streamlit expert, but do we need to generate this file even if user doesn't want to download it?

File: data/datasets/pl/graph/template_README.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+from huggingface_hub import hf_hub_download
+
+DATA_DIR = "<your_local_data_directory>"
+JSON_FILE = "data/judgment_graph.json"
+hf_hub_download(repo_id="JuDDGES/pl-court-graph", repo_type="dataset", filename=JSON_FILE, local_dir=DATA_DIR)
+
+with open(f"{DATA_DIR}/{JSON_FILE}") as file:
+    g_data = json.load(file)
+
+g = nx.node_link_graph(g_data)
```
Comment (asawczyn) [inline data/datasets/pl/graph/template_README.md]: To ensure that this code works, I suggest to create a test with it.

File: data/datasets/pl/graph/template_README.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            },
+            index_file,
+        )
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}({len(self)})"
+
+
+ds = PlCourtGraphDataset(root_dir="data/datasets/pyg")
+print(ds)
```
Comment (asawczyn) [inline data/datasets/pl/graph/template_README.md]: To ensure that this code works, I suggest to create a test with it.

File: juddges/data/pl_court_api.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+    @property
+    def schema(self) -> dict[str, list[str]]:
+        return {
+            "judgement": [
+                "_id",
+                "signature",
+                "date",
+                "publicationDate",
+                "lastUpdate",
```
Comment (asawczyn) [inline juddges/data/pl_court_api.py]: Do you know if this dict is initialized on every call? I am asking about low level implementation, I know that this function will be executed every time, but maybe implementation of Python is so sophisticated that it stores constant values somewhere.

 If yes, then I recommend cached_property.

File: juddges/evaluation/eval_structured_llm_judge.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+            field: self.compute_metrics(preds[field], golds[field])
+            for field in tqdm(golds.keys(), desc="Fields")
+        }
+
+    def compute_metrics(self, preds: list[str], golds: list[str]) -> dict[str, float]:
+        """Assesses single field prediction either by comparing raw string or using LLM-as-judge otherwise."""
+        assert len(golds) == len(preds)
+        llm_assessments = []
+        num_llm_evals = 0
+        enum_golds_preds = enumerate(zip(golds, preds))
```
Comment (asawczyn) [inline juddges/evaluation/eval_structured_llm_judge.py]:
```suggestion
        llm_assessments = []
        num_llm_evals = 0
        enum_golds_preds = enumerate(zip(golds, preds, strict=True))
```

File: juddges/evaluation/eval_structured_llm_judge.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+            try:
+                results_summary[judgement] += 1
+            except KeyError:
+                results_summary[INVALID_JUDGMENT] += 1
+
+        return {name: val / len(llm_assessments) for name, val in results_summary.items()}
+
+
+if __name__ == "__main__":
+    from dotenv import load_dotenv
```
Comment (asawczyn) [inline juddges/evaluation/eval_structured_llm_judge.py]:
```suggestion

# Example usage:
if __name__ == "__main__":
    from dotenv import load_dotenv
```

File: juddges/evaluation/eval_structured_llm_judge.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        }
+
+    def compute_metrics(self, preds: list[str], golds: list[str]) -> dict[str, float]:
+        """Assesses single field prediction either by comparing raw string or using LLM-as-judge otherwise."""
+        assert len(golds) == len(preds)
+        llm_assessments = []
+        num_llm_evals = 0
+        enum_golds_preds = enumerate(zip(golds, preds))
+        for i, (ans_gold, ans_pred) in (
+            pbar := tqdm(enum_golds_preds, total=len(golds), leave=False, desc="Evaluating")
```
Comment (asawczyn) [inline juddges/evaluation/eval_structured_llm_judge.py]: pbar should be closed. It is easier to use it with `with`

File: juddges/evaluation/info_extraction.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+        for eval in (
+            pbar := tqdm(self.structured_evaluators, desc="Structured", disable=not self.verbose)
+        ):
+            pbar.set_postfix({"eval": eval.name})
+            metrics[eval.name] = eval.evaluate(preds=preds, golds=golds)
+
+        text_preds = [res["answer"] for res in results]
+        text_golds = [res["gold"] for res in results]
+        for eval in (
+            pbar := tqdm(self.full_text_evaluators, desc="Full text", disable=not self.verbose)
```
Comment (asawczyn) [inline juddges/evaluation/info_extraction.py]: pbar should be closed. It is easier to use it with `with`

File: juddges/prompts/__init__.py [hunk 14]
```diff
@@ -0,0 +1 @@
+__version__ = "0.0.1"
```
Comment (asawczyn) [inline juddges/prompts/__init__.py]: Do we need version in the prompts module?

File: juddges/settings.py [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    "gpt-4-1106-preview": 0.01 / 1000,
+    "gpt-4-0125-preview": 0.01 / 1000,
+    "gpt-3.5-turbo-1106": 0.001 / 1000,
+}
+
+LLM_TO_PRICE_COMPLETION = {
+    "gpt-4-1106-preview": 0.03 / 1000,
+    "gpt-4-0125-preview": 0.03 / 1000,
+    "gpt-3.5-turbo-1106": 0.002 / 1000,
+}
```
Comment (asawczyn) [inline juddges/settings.py:47]: probably outdated?

File: nbs/Data/00_prepare_instruction_dataset_for_ir.ipynb [hunk 16]
```diff
@@ -0,0 +1,946 @@
+{
```
Comment (asawczyn) [inline nbs/Data/00_prepare_instruction_dataset_for_ir.ipynb]: I didn't check notebooks. Do I have to?

File: scripts/README.md [hunk 17]
```diff
@@ -0,0 +1,111 @@
+# Dataset preparation
```
Comment (asawczyn) [inline scripts/README.md]: is this readme mentioned in main readme?

File: scripts/graph_use_case.py [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+from torch_geometric.data import InMemoryDataset, download_url
+from torch_geometric.transforms import BaseTransform
+
+
+class PlCourtGraphDataset(InMemoryDataset):
+    URL = (
+        "https://huggingface.co/datasets/JuDDGES/pl-court-graph/resolve/main/"
+        "data/pyg_judgement_graph.pt?download=true"
+    )
+
```
Comment (asawczyn) [inline scripts/graph_use_case.py]: copy code to tests to ensure that it is working

**Review by asawczyn — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (binkjakub) [reply Makefile]: Installation from this file was always pain, conda infinitely "resolves environment".

File: nbs/Data/00_prepare_instruction_dataset_for_ir.ipynb (hunk 16 above)
Comment (binkjakub) [reply nbs/Data/00_prepare_instruction_dataset_for_ir.ipynb]: No

File: README.md (hunk 2 above)
Comment (binkjakub) [reply README.md]: When we were starting, unloth supported only `3.10`, now we can switch to `3.11`. Fixed

File: README.md (hunk 4 above)
Comment (binkjakub) [reply README.md]: Fixed

File: README.md (hunk 5 above)
Comment (binkjakub) [reply README.md:65]: Fixed

File: dashboards/pages/02_🔍_Analyse_Extracted_Information.py (hunk 6 above)
Comment (binkjakub) [reply dashboards/pages/02_🔍_Analyse_Extracted_Information.py:51]: idk, that's not my code

File: juddges/data/pl_court_api.py (hunk 9 above)
Comment (binkjakub) [reply juddges/data/pl_court_api.py]: I guess it'll create it everytime. I changed to `cached_property`.

File: juddges/evaluation/eval_structured_llm_judge.py (hunk 10 above)
Comment (binkjakub) [reply juddges/evaluation/eval_structured_llm_judge.py]: Fixed

File: juddges/evaluation/eval_structured_llm_judge.py (hunk 11 above)
Comment (binkjakub) [reply juddges/evaluation/eval_structured_llm_judge.py]: Fixed

File: juddges/evaluation/eval_structured_llm_judge.py (hunk 12 above)
Comment (binkjakub) [reply juddges/evaluation/eval_structured_llm_judge.py]: Fixed

File: juddges/evaluation/info_extraction.py (hunk 13 above)
Comment (binkjakub) [reply juddges/evaluation/info_extraction.py]: Fixed

File: juddges/prompts/__init__.py (hunk 14 above)
Comment (binkjakub) [reply juddges/prompts/__init__.py]: Good point, it should've accidentally moved from top-level __init__. Fixed

File: juddges/settings.py (hunk 15 above)
Comment (binkjakub) [reply juddges/settings.py:47]: That's not mine part, ignore it for now

File: scripts/README.md (hunk 17 above)
Comment (binkjakub) [reply scripts/README.md]: Yes

File: data/datasets/pl/graph/template_README.md (hunk 8 above)
Comment (binkjakub) [reply data/datasets/pl/graph/template_README.md]: That's in `scripts/graph_use_case.py`. Wrote test runnig this script.

File: scripts/graph_use_case.py (hunk 18 above)
Comment (binkjakub) [reply scripts/graph_use_case.py]: Added, I'll check how it affect CI performance, because it has to download 1.5GB file

Comment (binkjakub) [reply scripts/graph_use_case.py]: Whole CI takes 6 min (3 min for unit testing). I think it's acceptable

File: data/datasets/pl/graph/template_README.md (hunk 7 above)
Comment (binkjakub) [reply data/datasets/pl/graph/template_README.md]: Created new script with this code and run in tests

**Review by asawczyn — APPROVED** [review]
